### ISSUE TITLE
run control-plane as non-root

### ISSUE THREAD

**Issue by neolit123** [kubernetes/kubeadm#2473]
KEP
https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane
k/e issue: https://github.com/kubernetes/enhancements/issues/2568

> This KEP proposes that the control-plane in kubeadm be run as non-root. If containers are running as root an escape from a container may result in the escalation to root in host. CVE-2019-5736 is an example of a container escape vulnerability that can be mitigated by running containers/pods as non-root.

ALPHA 1.22:
- [ ] seccomp = runtime/default:
https://github.com/kubernetes/kubernetes/pull/100234
- [ ] e2e tests
TODO
- [ ] docs (optional for alpha)
TODO

**Comment by neolit123** [comment]
cc @vinayakankugoyal

**Comment by vinayakankugoyal** [comment]
Can we update the 
add feature gate: link in the description above to https://github.com/kubernetes/kubernetes/pull/102158

**Comment by neolit123** [comment]
@vinayakankugoyal https://kubernetes.io/blog/2019/01/15/container-storage-interface-ga/#how-to-use-a-csi-volume
if the rootless kubeadm apiserver eventually becomes ON by default, would it break CSI driver users?

**Comment by vinayakankugoyal** [comment]
no because it is not the kube-apiserver that needs to run as privileged pod, it is the csi driver that needs to run as privileged pod. --allow-privileged=true allows privileged containers it does not make kube-apiserver's container privileged. (Same for kubelet but that is anyways out of scope of this KEP.)

**Comment by vinayakankugoyal** [comment]
> @vinayakankugoyal https://kubernetes.io/blog/2019/01/15/container-storage-interface-ga/#how-to-use-a-csi-volume
> if the rootless kubeadm apiserver eventually becomes ON by default, would it break CSI driver users?

no because it is not the kube-apiserver that needs to run as privileged pod, it is the csi driver that needs to run as privileged pod. --allow-privileged=true allows privileged containers it does not make kube-apiserver's container privileged. (Same for kubelet but that is anyways out of scope of this KEP.)

**Comment by vinayakankugoyal** [comment]
Can we update e2e section above with PR: https://github.com/kubernetes/kubeadm/pull/2511

**Comment by pacoxu** [comment]
I am doing some investigations and testing on this alpha feature recently. 
Is there anything that should be done in 1.23? @neolit123  @vinayakankugoyal

**Comment by neolit123** [comment]
> I am doing some investigations and testing on this alpha feature recently.

let us know if you find any bugs. my biggest concern is around supporting linux distros that are non-standard in terms of system files.

> Is there anything that should be done in 1.23?

the KEP was not updated for 1.23, with the premise to give the alpha one more release for users to test it.
i didn't see anyone object to this plan.

**Comment by pacoxu** [comment]
It works well in my basic testing and will keep running such a non-root env to see if there is an issue.

**Comment by neolit123** [comment]
hey @vinayakankugoyal i saw you PR to turn on the FG by default:
https://github.com/kubernetes/kubernetes/pull/106869

in the issue description here i've enumerated the steps for this to graduate to beta in 1.24.
would you be able to work on these tasks in the next 4 months?

also note that starting next week i will be on PTO until early Jan 2022, so not sure how much i can review until then.

**Comment by vinayakankugoyal** [comment]
Hi @neolit123 (long time 😄 ). Thanks for updating the bug with the beta graduation work. Ill we able to work on these tasks in the next 4 months.

**Comment by neolit123** [comment]
as noted earlier, there seem to be some activity on supporting user namespaces for pods in core k8s:
https://github.com/kubernetes/enhancements/pull/3065
https://github.com/kubernetes/enhancements/pull/2101
(not sure which one is the KEP PR to watch, possibly the newer one).

> The goal of supporting user namespaces in Kubernetes is to be able to run
processes in pods with a different user and group IDs than in the host.
Specifically, a privileged process in the pod runs as an unprivileged process in
the host. If such a process is able to break out of the container to the host,
it'll have limited impact as it'll be running as an unprivileged user there.

@vinayakankugoyal what is your evaluation of the user namespaces KEP?
do you see it as something that has end-goal overlap with the kubeadm RootlessControlPlane FG?
you've mentioned that it would not support hostPath mounts for the Alpha. anything else to note about it?

user namespace support is a much desired change in k8s, and i consider what we have in kubeadm a bit of a hack that may bite us due to distro specific drift - we manually manage the user/groups to simplify the UX and so that users that want to not run the CP as root can get it automatically.

i think we need to evaluate whether we want to put a hold on the kubeadm feature moving to beta and instead waiting on the username spaces feature to go Beta, at which point we can start using it and set the right fields in the Pod spec and potentially remove the kubeadm feature.

but....the user namespaces KEP is still in review and there are some pending concerns and a lot of discussion there.
as we discussed with @fabriziopandini in today's kubeadm meeting, we would have to evaluate if that KEP is not going to move forward in time. if it moves forward nicely we might want to start using it at some point. in the meantime users can use the kubeadm alpha feature. if it does not move forward in time, we are going to graduate the kubeadm feature.

**Comment by pacoxu** [comment]
> the user namespaces KEP is still in review and there are some pending concerns and a lot of discussions there.
> as we discussed with @fabriziopandini in today's kubeadm meeting

IMO, there are some overlaps between RootlessControlPlane in kubeadm and user namespace support in kubelet. Just two ways to make it. However, there are no conflicts.

As RootlessControlPlane is alpha, could we re-design it to use user namespace if possible? It means that kubeadm can fall back to the current solution if the UserNamespace is not enabled on the master node.

We can promote RootlessControlPlane to beta in 1.24 as this is a good-enough solution in my mind.  If RootlessControlPlane keeps being alpha, it will benefit fewer users as it is not by default behavior. 
- If UserNamespaceRemapping will be alpha in 1.24, we can support migration to the user namespace solution. `kubeadm init`  will check UserNamespaceRemapping feature gate at first.  kubeadm can fall back to the current solution if the UserNamespace is not enabled on the master node.
- If UserNamespaceRemapping is not ready in 1.24, we can support migration to the user namespace solution once it is ready. RootlessControlPlane will keep being Beta.
- Meanwhile, we have to propone the graduation of kubeadm FG RootlessControlPlane util user namespace is beta or GA(Then we can remove RootlessControlPlane).

**Comment by neolit123** [comment]
> As RootlessControlPlane is alpha, could we re-design it to use user namespace if possible?

We can talk more about that. I have not seen similar FG redesigns in k8s, but sounds doable if we redesign the alpha. If our FG is already beta a redesign contradicts with the beta definition, at least in my book.

> We can promote RootlessControlPlane to beta in 1.24 as this is a good-enough solution in my mind. If RootlessControlPlane keeps being alpha, it will benefit fewer users as it is not by default behavior.

The main problem with promotion of FGs that we are not sure about to beta, is that users start enabling them in production even if the FG is off by default. Then they start binding their infra to the implementation in weird ways. E.g. reusing the kubeadm managed uid/gids.

I think if we are not sure about something, it is wiser to just wait...maybe one more release. I like where the conversations are going in the user namespace kep and i think they will solve the volume problems as well at some point. If we can, i think we should help drive that kep with what we can - reviews etc.

**Comment by pacoxu** [comment]
> Then they start binding their infra to the implementation in weird ways. E.g. reusing the kubeadm managed uid/gids. Beta also means - the underlying design is stable and ready to mature.

It seems that RootlessControlPlane is not a mature way. I walk through the KEP RootlessControlPlane and the risk https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane#risks-and-mitigations is still there. 

If this is not a mature way, it should keep alpha and be removed once UserNamespace(a better solution) is out.

All my suggestions in my last comments are based on that "either is a good-enough solution for non-root on nodes". If not, the redesign is not acceptable.

**Comment by pacoxu** [comment]
https://github.com/kubernetes/enhancements/pull/3065 is merged. 👍

**Comment by neolit123** [comment]
our e2e for this feature started failing yesterday. i have no explanation for the time being. but i don't think it's a kubeadm problem, so maybe something in core?
https://github.com/kubernetes/kubeadm/issues/2750

**Comment by pacoxu** [comment]
> our e2e for this feature started failing yesterday. i have no explanation for the time being. but i don't think it's a kubeadm problem, so maybe something in core? #2750

https://github.com/kubernetes/kubernetes/pull/113548(merged) may fix it. (a revert of https://github.com/kubernetes/kubernetes/pull/113408 that was merged hours before that. )

**Comment by neolit123** [comment]
it looks like the job has been green for a while, so maybe something else fixed it. the failures were in late august. i completely forgot about this..

https://k8s-testgrid.appspot.com/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-rootless-latest

**Comment by pacoxu** [comment]
I opened the test grid(You post months ago) and find it failed yesterday(😓). 

> [kubernetes/kubernetes#113548](https://github.com/kubernetes/kubernetes/pull/113548) may fix it. (a revert of [kubernetes/kubernetes#113408](https://github.com/kubernetes/kubernetes/pull/113408) that was merged hours before that. )

Yesterday's failure is caused by that.  Not failures in August.😄

**Comment by sftim** [comment]
Is this actually important-longterm? It's been a few years.

**Comment by pacoxu** [comment]
/remove-priority important-soon

> Is this actually important-longterm? It's been a few years.

This feature is an alternative way for the user namespace feature. As we prefer to use the user namespace to gain the security control plane in the future, we decided to not promote this one to beta. But we should keep this FG until user namespace https://github.com/kubernetes/enhancements/issues/127 is beta.

**Comment by pacoxu** [comment]
https://github.com/kubernetes/enhancements/issues/127
User Namespace is beta in v1.30. We may start the deprecation of RootlessControlPlane in v1.31.

**Comment by LyKos4** [comment]
Is this expected to be completed? For which pods it is expected to change the user to non root?

**Comment by neolit123** [comment]
> Is this expected to be completed? For which pods it is expected to change the user to non root

this feature is alpha and deprecated. please UserNamespaces instead:
https://github.com/kubernetes/enhancements/issues/127

once UserNamespaces becomes GA kubeadm will enable it by default.

**Comment by neolit123** [comment]
> on hold until further notice. we are waiting the user namespaces KEP to go GA:
> - https://github.com/kubernetes/enhancements/issues/127

stable is planned for 1.35
https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/127-user-namespaces/kep.yaml#L23

**Comment by neolit123** [comment]
just to note, hostNetwork for kubeadm at this stage is more of a backwards compatibility artifact. the more i played with https://github.com/kubernetes/kubernetes/pull/134402 i am convinced it was originally added only because it's simpler (lazier) than using hostPorts. it has some minor security hardening implications where something can listen explicitly on localhost instead of a node IP, but everything we have is behind TLS, anyhow.

what i mean by backwards compatibility here is that someone might rely on hostNetwork's behavior already it i don't think we can change the default to be hostNetwork=false at this point. so if for some reason we decided to make add a knob in kubeadm v1's API to enable rootless CP pods like in https://github.com/kubernetes/kubernetes/pull/134402 i don't think that can ever be the default behavior at this point.

our dilemma here is what to do with the various options that we have, eventually having hostNetwork=false + hostUsers=true as per your plans sounds ideal, but that might take years. on the other hand, this hostPort experiment requires no ecosystem changes, but it changes some of the component default listen addresses. also i haven't tested multi node yet as kinder/kind are failing for some reason, but i think it should work.

**Comment by HirazawaUi** [comment]
I believe that "Allow hostNetwork pods to use user namespaces" can provide an additional option for kubeadm, making it one of the beneficiaries. However, kubeadm is not the only one that can benefit from this.

For kubeadm, a significant number of users may currently rely on static pods using host network. We can provide an option to inform users: if they need to enable user namespaces, we will default to using hostPort. However, this might cause some disruptions to their environment (I suspect).

We can offer two options:

  * Users can retain the current behavior of hostNetwork and wait until the new KEP reaches the beta stage (I don’t expect it to take too long, as I am currently debugging this feature). At that point, they can start using the user namespace functionality.
  
  * Users can opt for hostPort, which would allow them to use user namespaces directly. However, this may introduce some breaking changes.

**Comment by HirazawaUi** [comment]
> for option 3 we removed the support prematurely...after the Chinese holidays are over we'd have to decide if we want to revert the FG.

Perhaps removing it prematurely wasn't a bad thing. If we hadn't removed the `RootlessControlPlane` feature gate, we wouldn't have discovered the various issues with the user namespace :)

> if option 1 is approved and can work in the long run that would be great.

I will strive to advance it for v1.35, and it currently lacks a PRR reviewer.

**Comment by neolit123** [comment]
> > if option 1 is approved and can work in the long run that would be great.
> 
> I will strive to advance it for v1.35, and it currently lacks a PRR reviewer.

the PR for the KEP LGTM, i can comment later there if needed too:
- https://github.com/kubernetes/enhancements/pull/5608

another issue is that we don't enabled core k8s features in kubeadm until they are GA, and given there will be a new core FG to disable the validation that would mean kubeadm won't be able to use the feature for a while. we could document a guide on how users can opt-in though.
